src/bin/ion/commands/generate/model.rs

```diff
@@ -58,19 +73,120 @@ impl DataModelNode {
     // Represents fully qualified name of the type
     // e.g. In Java, `org.example.Foo`
     //      In Rust, `crate::org::example::Foo`
-    type_name: FullyQualifiedTypeName,
+    pub(crate) type_name: FullyQualifiedTypeName,
     // For types with parameters this will represent the nested parameters
-    parameters: Vec<FullyQualifiedTypeReference>,
+    pub(crate) parameters: Vec<FullyQualifiedTypeReference>,
+}
+
+impl From<FullyQualifiedTypeName> for FullyQualifiedTypeReference {
+    fn from(value: FullyQualifiedTypeName) -> Self {
+        Self {
+            type_name: value,
+            parameters: vec![],
+        }
+    }
+}
+
+impl Display for FullyQualifiedTypeReference {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        if self.parameters.is_empty() {
+            return write!(f, "{}", self.type_name.join("."));
+        }
+        write!(f, "{}<", self.type_name.join("."))?;
+
+        for (i, parameter) in self.parameters.iter().enumerate() {
+            if i == self.parameters.len() - 1 {
+                write!(f, "{}", parameter)?;
+            } else {
+                write!(f, "{},", parameter)?;
+            }
+        }
+        write!(f, ">")
+    }
+}
+
+// This is useful for code generator to convert input `serde_json::Value` coming from tera(template engine) into `FullyQualifiedTypeReference`
+impl TryFrom<&Value> for FullyQualifiedTypeReference {
+    type Error = tera::Error;
+
+    fn try_from(v: &Value) -> Result<Self, Self::Error> {
+        let obj = v.as_object().ok_or(tera::Error::msg(
+            "Tera value can not be converted to an object",
+        ))?;
+        let mut type_name = vec![];
+        let mut parameters: Vec<FullyQualifiedTypeReference> = vec![];
+        for (key, value) in obj {
+            if key == "type_name" {
+                type_name = value
+                    .as_array()
+                    .unwrap()
+                    .iter()
+                    .map(|s| s.as_str().unwrap().to_string())
+                    .collect();
+            } else {
+                let parameters_result: Result<Vec<FullyQualifiedTypeReference>, tera::Error> =
+                    value
+                        .as_array()
+                        .unwrap()
+                        .iter()
+                        .map(|v| v.try_into())
+                        .collect();
+                parameters = parameters_result?;
+            }
+        }
+        Ok(FullyQualifiedTypeReference {
+            type_name,
+            parameters,
+        })
+    }
+}
+
+impl FullyQualifiedTypeReference {
+    pub fn with_parameters(&mut self, parameters: Vec<FullyQualifiedTypeReference>) {
+        self.parameters = parameters;
+    }
+}
+
+/// Builder for abstract data types
+pub enum AbstractDataTypeBuilder {
+    WrappedScalar(WrappedScalarBuilder),
+    Scalar(ScalarBuilder),
+    Sequence(SequenceBuilder),
+    Structure(StructureBuilder),
+}
```

Comment:
Is this necessary? Can you use the derive builder crate to do this instead?

Reply:
Removed this instead now have will create different methods like `build_structure_from_constraints`, `build_sequence_from_constraints` etc.